Drop notary-less code

cthulhu-rider opened the pull request.

cthulhu-rider committed:
go.mod: Upgrade dependencies

Upgrade to the latest revisions of `neo-go`, `pkg/interop` and `testify`
modules.

Signed-off-by: Leonard Lyubich <leonard@morphbits.io>

cthulhu-rider committed:
Stop support for notary-less mode and prohibit updates to it

Make update routine to panic if `notaryDisabled` flag is set. Remove all
notary-less artifacts from the contract (storage and methods) on update.
Prevent from updating during pending votes. Make GAS transfers on
Alphabet contract's update.

Signed-off-by: Leonard Lyubich <leonard@morphbits.io>

cthulhu-rider committed:
Provide functionality to dump NeoFS contracts locally

Add `tests/dump` package with `Creator` and `Reader` types which provide
I/O operations of contract states (including storages).

Signed-off-by: Leonard Lyubich <leonard@morphbits.io>

cthulhu-rider committed:
Provide command to dump NeoFS contracts from the remote network

Add `cmd/dump` command which accepts Neo RPC endpoint of the NeoFS
network and dumps NeoFS contracts into `testdata` directory.

Signed-off-by: Leonard Lyubich <leonard@morphbits.io>

cthulhu-rider committed:
Provide test framework for contracts' migration

Signed-off-by: Leonard Lyubich <leonard@morphbits.io>

cthulhu-rider committed:
Test migration of the contracts to notary mode

Signed-off-by: Leonard Lyubich <leonard@morphbits.io>

cthulhu-rider committed:
Dump current data of the contract from NeoFS TestNet and MainNet

Signed-off-by: Leonard Lyubich <leonard@morphbits.io>

cthulhu-rider committed:
common: Bump contract versions to v0.17.0

Signed-off-by: Leonard Lyubich <leonard@morphbits.io>

cthulhu-rider commented:
@roman-khimov did u mean in https://github.com/nspcc-dev/neofs-contract/pull/324#discussion_r1102903590 that we can call `AddPeerIR` method from the Inner Ring only and `AddPeer` one from storage nodes only?

Btw unit test fails due to changes in method.

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +78,143 @@ func _deploy(data interface{}, isUpdate bool) {
+		if len(records) == 0 {
+			panic("did not find a record of the Proxy contract in the NNS")
+		} else if records[0] != address.FromHash160(proxyContract) {
+			const hexTable = "0123456789abcdef"
```
`std.Itoa` for `proxyContract`?

cthulhu-rider replied:
I don't realize what should I do with `std.Itoa`

roman-khimov replied:
You can convert `proxyContract` to hex this way (`Itoa` has `base` parameter) and compare strings, it's just easier to do so.

cthulhu-rider replied:
then how can I convert `interop.Hash160` contract address to `int`? In 4B chunks?

roman-khimov replied:
`convert.ToInteger()`, NeoVM `int` is a not a Go `int`.

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +78,143 @@ func _deploy(data interface{}, isUpdate bool) {
+			const hexTable = "0123456789abcdef"
+			j := 0
+			for _, v := range proxyContract {
+				if records[0][j] != hexTable[v>>4] || records[0][j+1] != hexTable[v&0x0f] {
```
BE/LE? Stingified hashes use a different byte order usually.

cthulhu-rider replied:
So what endianness? Or u sggest to check both?

roman-khimov replied:
Strings use LE as per NeoGo notation.

cthulhu-rider replied:
reversed order

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +78,143 @@ func _deploy(data interface{}, isUpdate bool) {
+
+		// distribute 75% of available GAS:
+		//  - 50% to Proxy contract
+		//  - the rest is evenly distributed between Inner Ring and storage nodes
```
This comment probably belongs the the next hunk.

cthulhu-rider replied:
It summarizes what's going on

roman-khimov replied:
I mean it seems like it should placed after the subsequent hunk (because it does something not related to GAS distribution).

cthulhu-rider replied:
IMO this is the best place for it since everything below is related to GAS distribution, for example, we need netmap contract to calculated GAS amount only

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +78,143 @@ func _deploy(data interface{}, isUpdate bool) {
+		//  - 50% to Proxy contract
+		//  - the rest is evenly distributed between Inner Ring and storage nodes
+		netmapContract := args[1].(interop.Hash160)
+		if len(netmapContract) > 0 {
```
So it's optional, while `proxyContract` is not?

cthulhu-rider replied:
yes, do u think netmap should also be required?

roman-khimov replied:
I'd expect some symmetry there.

cthulhu-rider replied:
lets then just get proxy contract from NNS and ignore the argument

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +78,143 @@ func _deploy(data interface{}, isUpdate bool) {
+		})
+		innerRingNodes := common.InnerRingNodesFromNetmap(netmapContract)
+
+		currentContact := runtime.GetExecutingScriptHash()
```
`currentContract`?

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +78,143 @@ func _deploy(data interface{}, isUpdate bool) {
+			nil, // receiver account (set in loop)
+			0,   // till (defaults to current chain height + 5760)
+		}
+		// TODO: is till OK?
```
Theoretically yes (nodes should take care of it), but practically we had a number of problems with this, so we better push it to at least 6 months forward.

cthulhu-rider replied:
ok, let's assume that the blocks are always in 1s

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +78,143 @@ func _deploy(data interface{}, isUpdate bool) {
+		toTransfer = currentGAS - toTransfer
+
+		nNodes := len(storageNodes) + len(innerRingNodes)
+		perNodeGAS := toTransfer / nNodes
```
`/2`? You actually do two transfers per node below.

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +78,143 @@ func _deploy(data interface{}, isUpdate bool) {
+			}
+
+			notaryTransferData[0] = addr
+			if !gas.Transfer(currentContact, interop.Hash160(notary.Hash), perNodeGAS, notaryTransferData) {
```
Theoretically we could make notary deposits lower (like some fixed 20 GAS value).

cthulhu-rider replied:
Added limit

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +78,143 @@ func _deploy(data interface{}, isUpdate bool) {
+		}
+		// TODO: is till OK?
+
+		for i := 0; toTransfer > 0 && i < len(innerRingNodes); i++ {
```
Do we need to check `toTransfer`? `gas.Transfer` will fail anyway if it depletes prematurely.

cthulhu-rider replied:
if it is a good practice, then let's just transfer

roman-khimov commented on cmd/dump/blockchain.go:
```diff
@@ -0,0 +1,169 @@
+		return res, fmt.Errorf("call '%s' method of the NNS contract: %w", method, err)
+	}
+
+	records, err := unwrap.Array(callRes, nil)
```
Why not just `unwrap.Array(x.actor.Call(...))`?

cthulhu-rider replied:
because I added error context above to understand which method fell off

roman-khimov replied:
Well, it's not an error and can be left this way, but `unwrap.X()` were specifically designed for this use case, usually you get meaningful enough errors from it.

roman-khimov commented on cmd/dump/blockchain.go:
```diff
@@ -0,0 +1,169 @@
+		return res, fmt.Errorf("fetch stack item array from '%s' method's response of the NNS contract: %w", method, err)
+	}
+
+	// FIXME: temp hack
```
Fun one. But probably makes sense here.

cthulhu-rider replied:
then, if you do not mind, I would leave it. Code looks ugly, but `unwrap` functions do a lot of usefull stuff which I would not want to do here manually.

roman-khimov replied:
ok

roman-khimov commented on cmd/dump/main.go:
```diff
@@ -0,0 +1,98 @@
+
+func overtakeContracts(from *remoteBlockchain, to *dump.Creator) error {
+	for _, name := range []string{
+		"alphabet0",
```
Only one of them?

cthulhu-rider replied:
forgot to add TODO, anyway u caught this. Is `getcommittee` the most correct way to get Alphabet size?

roman-khimov replied:
It works for the purpose, yes.

roman-khimov commented on tests/dump/reader.go:
```diff
@@ -0,0 +1,105 @@
+
+// IterateContractStates iterates over all contracts from the superior dump and
+// passes their states into f.
+func (x *Reader) IterateContractStates(f func(name string, _state state.Contract)) (err error) {
```
Not that it's a big problem, but I'm not sure why you're wrapping `contracts` this way. You could just initialize them once (by decoding JSON) at `initDumpStreams` and then let them be accessible as a simple `Reader` field for whatever purpose. Now if I'm to require some additional passes over contracts I either have to copy them or call this method twice and decode JSON twice.

cthulhu-rider replied:
agree with pre-read, but I would not reveal the map, let it remain under the hood

roman-khimov commented on tests/dump/reader.go:
```diff
@@ -0,0 +1,105 @@
+
+// IterateContractStorages iterates over all contracts from the superior dump
+// and passes their storage items into f.
+func (x *Reader) IterateContractStorages(f func(name string, key, value []byte)) error {
```
Same thing here. Yes, it can be large, but we have to fit it into memory for the test anyway, so some `map[string][]kv` may just be easier to handle.

cthulhu-rider replied:
imo method is cleaner than direct map access, map provides to many knives. But I don't mind

roman-khimov commented on audit/audit_contract.go:
```diff
@@ -36,39 +36,64 @@ func (a auditHeader) ID() []byte {
 	}
 
 	args := data.(struct {
 		notaryDisabled bool
```
If that's the only thing left in `args`, just drop them completely. This won't affect compatibility, but will simplify code a bit.

roman-khimov commented on common/version.go:
```diff
@@ -4,23 +4,11 @@ import "github.com/nspcc-dev/neo-go/pkg/interop/native/std"
-
-	// Versions from which an update should be performed.
-	// These should be used in a group (so prevMinor can be equal to minor if there are
-	// any migration routines.
```
These are still relevant! We can't safely upgrade from 0.14.0, for example. At the same time IIRC we still have 0.15.x to 0.16.0 transition code.

cthulhu-rider replied:
> We can't safely upgrade from 0.14.0

Why? If there is a missing migration code, let's add it. Otherwise I can't test upgrade from version `<v0.16`. If we leave step-by-step update, then test framework should skip contracts `<common.PrevVersion`.

roman-khimov replied:
> there is a missing migration code

That's the case. But we don't have any pre-0.15.x networks in production and the upgrade procedure up to 0.16.0 was strictly sequential, so we may forget about pre-0.15.x at least. At the same time we still need this check, because we won't keep transition code forever. In a year we'll start dropping these old migrations and that's where we'll be bumping the minimal upgrade version.

roman-khimov commented on reputation/reputation_contract.go:
```diff
@@ -23,24 +22,64 @@ func _deploy(data interface{}, isUpdate bool) {
 	}
 
 	args := data.(struct {
 		notaryDisabled bool
```
Just drop it entirely.

roman-khimov commented on subnet/subnet_contract.go:
```diff
@@ -60,16 +59,60 @@ const (
 	}
 
 	args := data.(struct {
 		notaryDisabled bool
```
And here as well.

cthulhu-rider committed:
Prevent potential contract rollbacks

In previous implementation `common.CheckVersion` allowed contract to be
updated to earlier versions. This could be a problem since currently
forward-only updates are supported.

Make `common.CheckVersion` to panic if original contract version is less
than `common.Version` one.

Signed-off-by: Leonard Lyubich <leonard@morphbits.io>

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +80,148 @@ func _deploy(data interface{}, isUpdate bool) {
+		perNodeGASSimple := perNodeGAS - perNodeGASNotary
+
+		// see https://github.com/nspcc-dev/neo-go/blob/v0.101.0/docs/notary.md#1-notary-deposit
+		const lockInterval = 6 * 30 * 24 * 60 * 60 // 6 months blocks of 1s
```
Production networks never use `1s` for block time.

cthulhu-rider replied:
Can I access block interval from the contract somehow other than getting current and previous block timestamps?

roman-khimov replied:
No, and current-previous is not reliable either. You can safely assume 15s here, public networks use that.

roman-khimov commented on alphabet/alphabet_contract.go:
```diff
@@ -60,14 +80,148 @@ func _deploy(data interface{}, isUpdate bool) {
+		const lockInterval = 6 * 30 * 24 * 60 * 60 // 6 months blocks of 1s
+		notaryTransferData := []interface{}{
+			nil,                              // receiver account (set in loop)
+			runtime.GetTime() + lockInterval, // till
```
`GetTime()` is about UTC time, while we need blocks here (current height).

cthulhu-rider replied:
Misinterpreted docs, fixed

cthulhu-rider committed:
Fix `unused` linter's remark

`defaultRegisterPrice` constant is used in `_deploy` function of the NNS
contract.

Signed-off-by: Leonard Lyubich <leonard@morphbits.io>

cthulhu-rider commented:
* https://github.com/nspcc-dev/neofs-contract/pull/324
